packages/contracts-bedrock/test/properties/medusa/ProtocolAtomic.t.sol

```diff
@@ -0,0 +1,180 @@
+
+    function fuzz_DeployNewSupertoken(
+        TokenDeployParams memory params,
+        uint256 chainId
+    )
+        external
+        validateTokenDeployParams(params)
+    {
+        chainId = bound(chainId, 0, MAX_CHAINS - 1);
+        _deploySupertoken(
+            remoteTokens[params.remoteTokenIndex],
+            WORDS[params.nameIndex],
+            WORDS[params.symbolIndex],
+            DECIMALS[params.decimalsIndex],
+            chainId
+        );
+    }
+
+    /// @custom:property-id 22
+    /// @custom:property sendERC20 decreases sender balance in source chain and increases receiver balance in
+    /// destination chain exactly by the input amount
+    /// @custom:property-id 23
+    /// @custom:property sendERC20 decreases total supply in source chain and increases it in destination chain exactly
+    /// by the input amount
+    function fuzz_SelfBridgeSupertoken(uint256 fromIndex, uint256 destinationChainId, uint256 amount) external {
+        destinationChainId = bound(destinationChainId, 0, MAX_CHAINS - 1);
+        fromIndex = bound(fromIndex, 0, allSuperTokens.length - 1);
+        OptimismSuperchainERC20 sourceToken = OptimismSuperchainERC20(allSuperTokens[fromIndex]);
+        OptimismSuperchainERC20 destinationToken =
+            MESSENGER.crossChainMessageReceiver(address(sourceToken), destinationChainId);
+        // TODO: when implementing non-atomic bridging, allow for the token to
+        // not yet be deployed and funds be recovered afterwards.
+        require(address(destinationToken) != address(0));
+        uint256 sourceBalanceBefore = sourceToken.balanceOf(msg.sender);
+        uint256 sourceSupplyBefore = sourceToken.totalSupply();
+        uint256 destinationBalanceBefore = destinationToken.balanceOf(msg.sender);
+        uint256 destinationSupplyBefore = destinationToken.totalSupply();
+
+        vm.prank(msg.sender);
+        try sourceToken.sendERC20(msg.sender, amount, destinationChainId) {
+            uint256 sourceBalanceAfter = sourceToken.balanceOf(msg.sender);
+            uint256 destinationBalanceAfter = destinationToken.balanceOf(msg.sender);
+            // no free mint
+            assert(sourceBalanceBefore + destinationBalanceBefore == sourceBalanceAfter + destinationBalanceAfter);
+            // 22
+            assert(sourceBalanceBefore - amount == sourceBalanceAfter);
+            assert(destinationBalanceBefore + amount == destinationBalanceAfter);
+            uint256 sourceSupplyAfter = sourceToken.totalSupply();
+            uint256 destinationSupplyAfter = destinationToken.totalSupply();
+            // 23
+            assert(sourceSupplyBefore - amount == sourceSupplyAfter);
+            assert(destinationSupplyBefore + amount == destinationSupplyAfter);
+        } catch {
+            assert(address(destinationToken) == address(sourceToken) || sourceBalanceBefore < amount);
+        }
+    }
+
+    function fuzz_MintSupertoken(uint256 index, uint96 amount) external {
```

Comment:
Missing NATSPEC?

Reply:
I think a good idea could be to split the function to have the properties test functions on one place, and the test-helpers functions in other.
Wdyt?

Reply:
good idea!